codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -187,16 +187,22 @@ public void operationComplete(ChannelFuture future) throws Exception {
                 }
                 ChannelFuture future = frameWriter.writeHeaders(ctx, streamId, headers, streamDependency,
                                                                 weight, exclusive, padding, endOfStream, promise);
-                // Synchronously set the headersSent flag to ensure that we do not subsequently write
-                // other headers containing pseudo-header fields.
-                stream.headersSent();
+                // Writing headers may fail during the encode state if they violate HPACK limits.
+                Throwable failureCause = future.cause();
```

Comment:
We make assumptions about what is being done in `FrameWriter#writeHeaders` and properties of the underlying transport. In this context the method is assumed to encode the headers and call `ctx.write()` in a synchronous manner. Also if a write fails we assume the transport is reliable/ordered and therefore unusable so we generally don't care about promise failures. So I'm not sure we need to handle the async failure case, and if you think we do this needs to be done more generally throughout codec-http2. WDYT?